buckets/local/bucket.go

```diff
@@ -0,0 +1,297 @@
+			}
+			p := n
+			n = strings.TrimPrefix(n, abs+"/")
+			if strings.HasPrefix(n, filepath.Dir(archiveName)+"/") || strings.HasSuffix(n, PatchExt) {
+				return nil
+			}
+			file, err := os.Open(p)
+			if err != nil {
+				return err
+			}
+			defer file.Close()
+			nd, err := addFile(b.tmp, b.layout, file)
+			if err != nil {
+				return err
+			}
+			if err = editor.InsertNodeAtPath(ctx, n, nd, unixfs.EmptyDirNode); err != nil {
+				return err
+			}
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	return editor.Finalize(ctx, b.tmp)
+}
+
+// createArchive creates the empty archive file under path.
+func (b *Bucket) createArchive() (*os.File, error) {
+	name := filepath.Join(b.path, archiveName)
+	if err := os.MkdirAll(filepath.Dir(name), os.ModePerm); err != nil {
+		return nil, err
+	}
+	return os.Create(name)
+}
+
+// ArchiveFile creates an archive describing a directory containing reader.
+func (b *Bucket) ArchiveFile(ctx context.Context, pth string, name string) error {
+	r, err := os.Open(pth)
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+	root := unixfs.EmptyDirNode()
+	prefix, err := md.PrefixForCidVersion(1)
+	if err != nil {
+		return err
+	}
+	root.SetCidBuilder(prefix)
+	editor := dagutils.NewDagEditor(root, b.tmp)
+	f, err := addFile(b.tmp, b.layout, r)
+	if err != nil {
+		return err
+	}
+	if err = editor.InsertNodeAtPath(ctx, name, f, unixfs.EmptyDirNode); err != nil {
+		return err
+	}
+	n, err := editor.Finalize(ctx, b.tmp)
+	if err != nil {
+		return err
+	}
+	file, err := b.createArchive()
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	if err = car.WriteCarWithWalker(ctx, b.tmp, []cid.Cid{n.Cid()}, file, carWalker); err != nil {
+		return err
+	}
+	b.root = n.Cid()
+	return nil
+}
+
+// HashFile returns the cid of the file at path.
+func (b *Bucket) HashFile(pth string) (cid.Cid, error) {
+	r, err := os.Open(pth)
+	if err != nil {
+		return cid.Undef, err
+	}
+	defer r.Close()
+	n, err := addFile(b.tmp, b.layout, r)
+	if err != nil {
+		return cid.Undef, err
+	}
+	return n.Cid(), nil
+}
+
+// Diff returns a list of changes that are present in pth compared to the current archive.
+func (b *Bucket) Diff(ctx context.Context, pth string) (diff []*dagutils.Change, err error) {
+	var an ipld.Node
+	if b.root.Defined() {
+		an, err = b.tmp.Get(ctx, b.root)
+		if err != nil {
+			return
+		}
+	} else {
+		an = unixfs.EmptyDirNode()
+	}
+	bn, err := b.walkPath(ctx, pth)
+	if err != nil {
+		return
+	}
+	return dagutils.Diff(ctx, b.tmp, an, bn)
+}
+
+// Ignore returns true if the path contains an ignored file.
+func Ignore(pth string) bool {
+	for _, n := range ignoredFilenames {
+		if strings.HasSuffix(pth, n) {
+			return true
+		}
+	}
+	return false
+}
+
+// addFile chunks reader with layout and adds blocks to the dag service.
+// SHA2-256 is used as the hash function and CidV1 as the cid version.
+func addFile(dag ipld.DAGService, layout options.Layout, r io.Reader) (ipld.Node, error) {
```

Comment:
This is the actual file adding... it will add a file in the exact same way as unixfs.

Reply:
We could have an option for using a local blockstore on disk. The downside there is the badger files can get huge, which would feel silly when the bucket has 1MB of data. Maybe you could run a command that flips to an on disk blockstore when the RAM usage downside starts being a problem for larger buckets.

Reply:
Well, the "badger problem" is a bit related to removing keys that doesn't get GCed properly (without the right config, which seem to be one that I found).
Despite that, yeah, we could take different decision reg the bucket size. I gues using flat-file datastore could be an option too (since "ds" usage its quite simple here, and ephemeral).

Reply:
Nice, sounds good. I'll create an issue.